Custom rollout depth

jancervenka opened the pull request:
Draft resolving https://github.com/JuliaPOMDP/MCTS.jl/issues/88

jancervenka committed:
custom rollout depth

BoZenKhaa commented:
Hi Honza! Thanks for a quick attempt at this! I have a few comments.

First, I like the approach of putting the depth as a parameter of the rollout estimator. I think this is elegant and easy to follow. Though as a consequence, it will be difficult to use the current depth in the tree in the rollouts. At least for me, this is not a big issue and it's simpler this way, but maybe in some use cases, this is needed. What do you think @zsunberg? This could be addressed by still passing the depth to the `estimate_value` function, and implementing more rollout estimators:

- default rollout estimator, that does not use depth passed to it but instead uses depth defined in `SolvedRolloutEstimator`
- depth-dependent estimator, that does use depth passed to `rollout` and replicates the old behavior

Regarding the code, I tried to add commits in your branch that will let the tests run, I haven't figured out how to add them here (do you know how to do that?).  

- https://github.com/jancervenka/MCTS.jl/pull/1 The `new` in the inner constructor of `SolvedRolloutEstimator` requires types:
```
    function SolvedRolloutEstimator(policy, rng, depth::Union{Int, Nothing}=nothing)
        new{typeof(policy), typeof(rng)}(policy, rng, depth)
    end
```
- https://github.com/jancervenka/MCTS.jl/pull/2  `POMDPs.RolloutSimulator` in `rollout` function needs to be called with keyword arguments since there is no method taking `nothing` as positional argument:
```
RolloutSimulator(rng::AbstractRNG, d::Int=typemax(Int)) #not possible to use nothing here for depth
RolloutSimulator(;rng=Random.GLOBAL_RNG, eps=nothing,   max_steps=nothing) # you can pass nothing to this method using kwargs
```

Also, you can test your code quickly from REPL. Some tests are failing right now, and this is how you run them:

1. launch REPL in the root of the MCTS package
2. turn package mode on in REPL with ']'
2. run `activate .` to activate the environment
3. run `test` to run the tests

Alternatively, you can run the `runtests.jl` file, but the tests may be using a different environment than the package. The above method should handle that.

I think as the next step, you can have a look at the test set, but let's wait for @zsunberg before fixing them.

jancervenka commented:
Thank you for the feedback! I will fix the tests once we agree that the proposed solution is OK.

Btw the best way to suggest changes in PR is to create a comment and use the "add a suggestion" feature.
<img width="1221" alt="Screenshot 2022-02-21 at 14 36 20" src="https://user-images.githubusercontent.com/22942125/154965852-3487906f-de0b-4df9-b9f2-c93a239ac4d6.png">

BoZenKhaa commented on src/domain_knowledge.jl:
```diff
@@ -31,6 +31,11 @@ This is within the policy when a RolloutEstimator is passed to an AbstractMCTSSo
+    depth::Union{Int, Nothing}
+
+    function SolvedRolloutEstimator(policy, rng, depth::Union{Int, Nothing}=nothing)
+        new(policy, rng, depth)
```
```suggestion
        new{typeof(policy), typeof(rng)}(policy, rng, depth)
```

BoZenKhaa commented on src/domain_knowledge.jl:
```diff
@@ -42,20 +47,20 @@ convert_to_policy(s::Solver, mdp::Union{POMDP,MDP}) = solve(s, mdp)
-function rollout(estimator::SolvedRolloutEstimator, mdp::MDP, s, d::Int)
-    sim = RolloutSimulator(estimator.rng, d)
+function rollout(estimator::SolvedRolloutEstimator, mdp::MDP, s)
+    sim = RolloutSimulator(estimator.rng, estimator.depth)
```
```suggestion
    sim = RolloutSimulator(rng=estimator.rng, max_steps=estimator.depth)
```

BoZenKhaa reviewed (commented):
These changes allow for runtest.jl to run.

jancervenka committed:
Apply suggestions from code review

Co-authored-by: Jan <mrkosjan@gmail.com>

zsunberg commented:
@WhiffleFish can you please review these changes and let me know what you think.

WhiffleFish commented on src/domain_knowledge.jl:
```diff
@@ -31,6 +31,11 @@ This is within the policy when a RolloutEstimator is passed to an AbstractMCTSSo
 mutable struct SolvedRolloutEstimator{P<:Policy, RNG<:AbstractRNG}
     policy::P
     rng::RNG
+    depth::Union{Int, Nothing}
```
The end-user doesn't directly work with `SolvedRolloutEstimator`. Rather, the user instantiates a `RolloutEstimator` object  with all necessary options and puts it in the `estimate_value` field for MCTS (or any other applicable solver).

I would recommend giving the `RolloutEstimator` struct a `max_depth` field that then fills the corresponding field in `SolvedRolloutEstimator` in the `convert_estimator` call.

BoZenKhaa replied:
I think `eps` could also be added, with default <1.

WhiffleFish replied:
You're right. I think a reasonable middle ground between infinite rollouts and a low default `max_depth` would be an epsilon field that defaults to something like 0.01. In this case, a pomdp with discount 0.99 would run for 459 steps which is high but not unreasonable.

WhiffleFish commented on src/domain_knowledge.jl:
```diff
@@ -31,6 +31,11 @@ This is within the policy when a RolloutEstimator is passed to an AbstractMCTSSo
     rng::RNG
+    depth::Union{Int, Nothing}
+
+    function SolvedRolloutEstimator(policy, rng, depth::Union{Int, Nothing}=nothing)
```
By default here, rollouts will run until a terminal state is hit. Provided that there are plenty POMDPs/MDPs without terminal states (tiger, crying baby, etc.) these defaults will fail silently with the rollout running in an infinite loop. Even if we do throw one warning deterring users from falling into this infinite loop trap, it would be an unnecessary nuisance.

I would recommend setting some default maximum depth or default epsilon that prevents this.

Also, given that the end-user doesn't usually directly interact with `SolvedRolloutEstimator`, this constructor may be unnecessary.

BoZenKhaa replied:
Thanks for having a look. I agree that cyclic/infinite horizon (PO)MDPs should be handled gracefully. 

I think the default rollout estimator with finite depth and eps<1 should be fine, passing these values to the `RolloutSimulator`. There will be a slight inconsistency since the defaults for `RolloutSimulator` are `nothing` for both values. But that's better than having the infinite loops I think.

I have a question though, what do you think about removing the `depth` value from the call to `estimate_value` in this PR? IMO its an improvement, but it is a significant change since it will be impossible to replicate the old rollout behavior (with `depth` parameter bounding _combined_ depth of tree _and_ rollout - with different rollouts being of different depth as a result). Do you know if the old behavior is important to anyone?

It is possible to keep the `depth` parameter in the calls to `estimate_value` and have the new `RolloutEstimator` handling the new way of rollouts, and then add a `LegacyRolloutEstimator` replicating the old behavior, but I think it would be better to remove this complexity if it is not needed.

WhiffleFish replied:
I absolutely believe this to be a worthwhile pursuit. I also totally forgot that `RolloutSimulator` has `nothing` as both defaults, however because people tend to use a random rollout simulator as the default value estimator (MCTS, POMCPOW, etc.) the depth is hidden behind a second layer of abstraction.

It certainly wouldn't be impossible to maintain the legacy functionality, however I don't see the need to do so given that the old approach has had some noted problems and doesn't seem to offer an strict benefits relative to this newer approach. In my opinion you're right in that it would be better to minimize complexity and change over to constant depth rollouts while dropping this old functionality.

WhiffleFish reviewed (changes requested):
Overall, I think the general approach looks good. However, in my opinion, it still requires a few more changes before we merge.

As a final note, some tests failed with the changes. Line 26 of `test/options.jl` errored because it still included depth as an argument in the value estimation call, so that should be rectified possibly along with others that may fail once this one doesn't error.

jancervenka committed:
review

jancervenka commented:
Thank you for the review! @BoZenKhaa @WhiffleFish 
- I have moved the `max_depth` attribute and the default constructor to `RolloutEstimator` and propagated the value to `SolvedRolloutEstimator`
- `max_depth` now has a finite default value of `max_depth=50`. Is it reasonable?
- I have added the `eps` attribute to `RolloutEstimator` so it exposes both the `max_steps` and `eps` of `RolloutSimulator` API. The default value is `eps=nothing`, is that okay or should I change it to `eps=0.01`?
- The tests are now passing

jancervenka committed:
depth as max_depth

BoZenKhaa commented on src/vanilla.jl:
```diff
@@ -25,8 +25,8 @@ Fields:
+        If this is a function `f`, `f(mdp, s)` will be called to estimate the value.
+        If this is an object `o`, `estimate_value(o, mdp, s)` will be called.
         If this is a number, the value will be set to that number
         default: RolloutEstimator(RandomSolver(rng))
```
I think the default here could be 
```
default: RolloutEstimator(RandomSolver(rng); max_depth=50, eps=nothing)
```
to highlight the change.

BoZenKhaa replied:
Also, please have a look at the docs as well so that they stay consistent: https://juliapomdp.github.io/MCTS.jl/latest/

BoZenKhaa commented on notebooks/Domain_Knowledge_Example.ipynb:
```diff
@@ -201,7 +201,7 @@
+    "function MCTS.estimate_value(h::MyHeuristic, mdp::SimpleGridWorld, s)\n",
     "    targ = h.target_state\n",
     "    m_dist = abs(s.x-targ.x)+abs(s.y-targ.y)\n",
     "    val = 10.0/m_dist\n",
```
On line 354, the max_depth could also be added to the call to RolloutEstimator.

jancervenka committed:
max_depth, eps as kwargs, docs update

jancervenka commented:
@BoZenKhaa I have added your suggestions. I have not found any inconsistencies in the documenttation.

BoZenKhaa commented:
> @BoZenKhaa I have added your suggestions. I have not found any inconsistencies in the documenttation.

I meant this
![image](https://user-images.githubusercontent.com/1710518/158762916-58dc2494-86a8-4f68-af76-b5ae44143571.png)
but it's most definitely minor thing. 

I very much like the state of the changes!

zsunberg commented:
Hi all, sorry for the delayed response - I have been quite busy here. The changes are quite nice. Thank you for being so thorough.

One question to consider here: In this PR you have reduced the arguments for `estimate_value` from `(mdp, s, d)` to `(mdp, s)`. I think that, instead, we should keep this argument (in case anyone wants to implement the old behavior), but just ignore it by default.

Any other thoughts?

zsunberg reviewed (changes requested):
In the docstrings for the solvers, please change the documentation for the `depth` argument to reflect that it no longer applies to the rollouts.

zsunberg commented:
@WhiffleFish are the changes you requested complete?

BoZenKhaa commented:
@zsunberg 
> One question to consider here: In this PR you have reduced the arguments for `estimate_value` from `(mdp, s, d)` to `(mdp, s)`. I think that, instead, we should keep this argument (in case anyone wants to implement the old behavior), but just ignore it by default.
> 
> Any other thoughts?

We discussed this with @WhiffleFish [in the review](https://github.com/JuliaPOMDP/MCTS.jl/pull/91/commits/07861d1e4260ca92d13de58c9c8c45bae31306bc#r824379419). The outcome was that making the new behavior default is a breaking change, keeping the legacy functionality would not be generally useful and would unnecessarily complicate things. 

If you agree, I think this is ready :-D

zsunberg commented:
Hi all, sorry that this fell stale - this semester was very busy for me, but it just ended. I think it is useful to *allow* the old behavior, so I will go through and add the depth argument back in and fix the other minor issues and merge this tomorrow.

zsunberg committed:
added back depth argument

zsunberg committed:
added test for max_depth=-1

zsunberg commented:
Thanks a bunch @jancervenka @BoZenKhaa ! I am sorry that this took so long, but it is a useful addition!
